### PR TITLE
Revamp the storage SPI to support new storage engine as an extension

### PR DESCRIPTION
**Pull request by chtyim**
- Added StorageProvider to storage SPI and implementations to support extension based storage engine.
- Removed StructuredTableRegistry from SPI as it is the implementation details of the current Table and PostgreSQL implementations.
- Changed the StructureTableAdmin API to replace the getSpecification method with exists and getSchema methods.
- Added method to check for compatibility between StructuredTableSchema and StructuredTableSpecification.
  - The schema of the actual table might be different than the specification in terms of data type. However, as long as the
    table can store data correctly without losing precision, it is considered as compatible.
- Refactored and code cleanup around StoreDefinition class and its usages.
- Unify StructuredTable related unit tests to use always Guice injector for creating StructuredTableAdmin and TransactionRunner.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/CreateSystemTablesStage.java [hunk 1]
```diff
@@ -43,17 +45,23 @@ public CreateSystemTablesStage(StructuredTableAdmin structuredTableAdmin) {
...
     for (StructuredTableSpecification spec : input.getSystemTables()) {
-      StructuredTableSpecification existing = structuredTableAdmin.getSpecification(spec.getTableId());
-      if (existing == null) {
-        // it's possible this throws TableAlreadyExistsException if two apps are deployed at the same time and there
-        // is a race. In that case, fail deployment. On re-deployment, the existing spec will get checked with the
+      try {
+        StructuredTableSchema schema = structuredTableAdmin.getSchema(spec.getTableId());
+
+        // If the table exists, check if the schema is compatible with the spec
+        if (!schema.isCompatible(spec)) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/CreateSystemTablesStage.java:52]: why make the change to only check schema compatibility?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/ApplicationDetailFetcherProvider.java [hunk 2]
```diff
@@ -48,11 +48,6 @@ public ApplicationDetailFetcher get() {
     if (storageImpl.equals(Constants.Dataset.DATA_STORAGE_NOSQL)) {
       return injector.getInstance(RemoteApplicationDetailFetcher.class);
     }
-    if (storageImpl.equals(Constants.Dataset.DATA_STORAGE_SQL)) {
-      return injector.getInstance(LocalApplicationDetailFetcher.class);
-    }
-    throw new UnsupportedOperationException(
-      String.format("%s is not a supported storage implementation, the supported implementations are %s and %s",
-                    storageImpl, Constants.Dataset.DATA_STORAGE_NOSQL, Constants.Dataset.DATA_STORAGE_SQL));
+    return injector.getInstance(LocalApplicationDetailFetcher.class);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/ApplicationDetailFetcherProvider.java:51]: not really part of this PR, but it seems like these classes are all the same, except for which class is remote and which is local. Should we take this opportunity to refactor into common class?

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java [hunk 3]
```diff
@@ -89,4 +96,61 @@ public boolean isIndexColumn(String fieldName) {
...
+  
+  /**
+   * Checks if this schema is compatible with the given {@link StructuredTableSpecification}. They are compatible if
+   *
+   * <ol>
+   *   <li>
+   *     This schema contains all the fields in the specification.
+   *   </li>
+   *   <li>
+   *     Each schema field has data type that can store the corresponding spec field data without losing precision.
```
Comment (albertshau) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java:132]: this doesn't end up being a problem in most implementations? Like if you try to write an int to a long column with jdbc, it doesn't error?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/CreateSystemTablesStage.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/pipeline/CreateSystemTablesStage.java:52]: ok I see further down compatibility check also checks primary key and indexes

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java (hunk 3 above)
Comment (chtyim) [reply cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java:132]: The StructuredTable implementation is the one who should handle it correctly if type conversion is needed. E.g. if a DB only has `long` type, but the `upsert` method is providing an `Integer` Field (a java type), then the StructuredTable implement needs to do the type widening if necessary before writing to DB. However, for JDBC, I think it is already handled in the Java type widening fashion.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/ApplicationDetailFetcherProvider.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/ApplicationDetailFetcherProvider.java:51]: Probably to do it in a different PR. The refactoring around those guice modules are not so obvious.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/config/SqlUserConfigStoreTest.java [hunk 4]
```diff
@@ -49,11 +56,21 @@ public static void setup() throws Exception {
     cConf.set(Constants.REQUIREMENTS_DATASET_TYPE_EXCLUDE, Joiner.on(",").join(Table.TYPE, KeyValueTable.TYPE));
 
     pg = PostgresInstantiator.createAndStart(cConf, TEMP_FOLDER.newFolder());
-    DataSource dataSource = pg.getPostgresDatabase();
-    SqlStructuredTableRegistry registry = new SqlStructuredTableRegistry(dataSource);
-    registry.initialize();
-    admin = new PostgresSqlStructuredTableAdmin(registry, dataSource);
-    TransactionRunner transactionRunner = new SqlTransactionRunner(admin, dataSource);
+    Injector injector = Guice.createInjector(
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/config/SqlUserConfigStoreTest.java:59]: This looks same for basically all sql related table unit tests, can this be refactored into a common place?

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java:132]: I am also curious about why this change is preferred. There might be some databases which forces on the actual type, i.e, not allow writing int type to long(though I haven't heard any). Will this change prevent the extension to be working?

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/field/FieldType.java [hunk 5]
```diff
@@ -32,12 +33,22 @@
...
+    INTEGER(Collections.emptySet()),
+    LONG(Collections.singleton(INTEGER)),
+    FLOAT(Collections.emptySet()),
+    DOUBLE(Collections.singleton(FLOAT)),
+    STRING(Collections.emptySet()),
+    BYTES(Collections.emptySet());
+
+    private final Set<Type> compatibleTypes;
+
+    Type(Set<Type> compatibleTypes) {
```
Comment (yaojiefeng) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/field/FieldType.java:45]: I am curious about why this change is preferred. There might be some databases which forces on the actual type, i.e, not allow writing int type to long(though I haven't heard any). Will this change prevent the extension to be working?

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/StructuredTableRegistry.java [hunk 6]
```diff
@@ -27,12 +29,6 @@
  */
 @Beta
 public interface StructuredTableRegistry {
```
Comment (yaojiefeng) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/StructuredTableRegistry.java:31]: This change moves this class out of SPI module, and also there is some changes in `StructuredTableAdmin` about get spec for the table, which makes the SPI backward incompatible. I know it is no big deal since there is no other storage implementation using the SPI now. 
Just wondering in the future, for the SPI changes, should we also deprecate first and remove on a major release?

**Review by yaojiefeng — COMMENTED** [review]
Just some minor questions

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java (hunk 3 above)
Comment (chtyim) [reply cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/StructuredTableSchema.java:132]: If the StructuredTableSpecification defined a field to be `INT` type, and yet the storage engine only supports `LONG`, then the StructuredTable implementation of that storage engine should perform the conversion when writing and reading the data. Hence it doesn't exist a case that trying to write a `int` to a `long` at the storage engine level.

As for the compatibility checks, it is an abstraction for schema compatibility in CDAP, regardless of the storage engine itself.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/StructuredTableRegistry.java (hunk 6 above)
Comment (chtyim) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/StructuredTableRegistry.java:31]: Currently all SPI implementations are inside cdap repo, also, these APIs are marked as `@Beta`. In future, once we removed the `@Beta` annotation, then we should be careful about the changes.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/config/SqlUserConfigStoreTest.java (hunk 4 above)
Comment (chtyim) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/config/SqlUserConfigStoreTest.java:59]: Currently these tests are scattered across different modules, making it kind of hard to refactor. Ideally we should move the postgreSQL implementation as an extension, and have tests defined there.

Open https://cdap.atlassian.net/browse/CDAP-18629 as a followup task.

**Review by yaojiefeng — APPROVED** [review]
LGTM
